src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/rollup/RollupRunner.kt

```diff
@@ -109,5 +130,331 @@ object RollupRunner : ScheduledJobRunner,
+            // If metadata does not exist, it will either be initialized for the first time or it will be recreated to communicate the failed state
+            if (rollupSearchService.shouldProcessRollup(job, metadata)) {
+                // Attempt to acquire lock
+                val lock = acquireLockForRollupJob(job, context)
+                if (lock == null) {
+                    logger.debug("Could not acquire lock for ${job.id}")
+                } else {
+                    runRollupJob(job, context, lock)
+                    // Release lock
+                    val released: Boolean = releaseLockForRollupJob(context, lock)
+                    if (!released) {
+                        logger.debug("Could not release lock for ${job.id}")
+                    }
+                }
+            } else if (job.isEnabled) {
+                // We are doing this outside of ShouldProcess as schedule job interval can be more frequent than rollup and we want to fail
+                // validation as soon as possible
+                when (val jobValidity = isJobValid(job)) {
+                    is RollupJobValidationResult.Invalid -> {
+                        val lock = acquireLockForRollupJob(job, context)
+                        if (lock != null) {
+                            setFailedMetadataAndDisableJob(job, jobValidity.reason)
+                            logger.info("updating metadata service to disable the job [${job.id}]")
+                            releaseLockForRollupJob(context, lock)
+                        }
+                    }
+                    else -> {}
+                }
+            }
+        }
+    }
+
+    private suspend fun acquireLockForRollupJob(rollupJob: ScheduledJobParameter, context: JobExecutionContext): LockModel? {
+        var lock: LockModel? = null
+        try {
+            // acquireLock will attempt to create the lock index if needed and then read/create a lock. This is purely for internal purposes
+            // and should not need the role's context to run
+            BackoffPolicy.exponentialBackoff(TimeValue.timeValueMillis(
+                RollupSettings.DEFAULT_ACQUIRE_LOCK_RETRY_DELAY),
+                RollupSettings.DEFAULT_ACQUIRE_LOCK_RETRY_COUNT
+            ).retry(logger) {
+                lock = context.lockService.suspendUntil { acquireLock(rollupJob, context, it) }
+            }
+        } catch (e: Exception) {
+            logger.error("Error trying to acquireLock", e)
+        }
+
+        return lock
+    }
+
+    private suspend fun releaseLockForRollupJob(context: JobExecutionContext, lock: LockModel): Boolean {
+        return context.lockService.suspendUntil { release(lock, it) }
+    }
+
+    // TODO: Clean up runner
+    // TODO: Scenario: The rollup job is finished, but I (the user) want to redo it all again
+    /*
+    * TODO situations:
+    *  There is a rollup.metadataID and doc but theres no job in target index?
```

Comment:
Just a heads up, I believe we do recover from this scenario now, however, it's not exactly as the TODO message expects:

* During `isJobValid`, we'll validate source and target index
* If we get the condition in this scenario, where rollup.metadataID and doc exist but there is no job in target index, we resolve it by updating the mappings for the target index and add the job in

So we recover from it but neither start over or move to `FAILED`, I suppose it would recover silently and continue from wherever the metadata has left off.

We can go over this behavior as part of reviewing retry logic later on to discuss if it's what we want or not.